### PR TITLE
Revamp prices handling, use latest prices library

### LINKED ISSUE #1487: Add support for prices 1.0.0-beta
(Please take the time and describe the problem you're having before you propose a solution.)

### What I'm trying to achieve

New version of prices, django-prices and django-prices-openexchangerates are published we should those new versions in saleor.

Saleor should use new api from:
prices >= 1.0.1-beta
django-prices >= 1.0.0-beta
django-prices-openexchangerates >= 1.0.0-beta

### PR DESCRIPTION
**Pull request by rafalp**
Fixes #1487

Loose list of things to get done before this PR may be mergeable:

* [x] Change past migrations to use `DecimalField` instead of `PriceField` - this is because API change in `PriceField` breaks old migrations.
* [x] Duct-tape around the code to make it run with updated libs and old price fields replaced with amount fields
* [x] Normalize money to `Price` before moving it around the app
* [x] Update to latest `prices` and `django-prices` versions
* [x] `django_prices_openexchangerates.exchange_currency` doesn't work with new API
* [x] Do something about `zero_amount = Money(0, currency=settings.DEFAULT_CURRENCY) zero=Price(zero_amount, zero_amount)` that litters the code
* [x] Make all imports for `Money`, `TaxedMoney` use `prices` instead of `django-prices`
* [x] Redo discounts (sales and vouchers) handling now that `prices` discounts are functions and not objects. (needs fix on order's recount on order view)
* [x] Add tests for refund/capture payment forms
* [x] Move `unit_price_net` and `unit_price_gross` fields in `OrderLine` to `TaxedMoneyField`
* [x] Explore changing `total_net` `total_tax` fields to `total_net` and `total_gross` for `TaxedMoneyField`
* [x] Cleanup FIXME bits in which we are casting `Money` to `TaxedMoney` via `price = TaxedMoney(price, price)` and such
* [x] Rebase against master
* [x] Document new approach to prices/values handling in our docs
* [x] Fix showstopper on when voucher's limit is entered in the dashboard
* [x] Consider moving `ZERO_TAXED_MONEY` to `core.utils`
* [x] Replace `shipping_price` field on order with `shipping_price_net`, `shipping_price_gross` and `shipping_price = TaxedMoneyField()`
* [x] Update GraphQL to work with new money types
* [x] Complete review changes
* [x] Squash and rebase

As we haven't settled yet on `prices` final API, this PR's also an occasion to test how new things will actually work when used. So pausing this work to make further changes downstream is on the table.

### Pull Request Checklist

(Please keep this section. It will make maintainer's life easier.)

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] Python code quality checks pass: `pycodestyle`, `pydocstyle`, `pylint`.
1. [x] JavaScript code quality checks pass: `eslint`.

### CODE REVIEW

**Comment by rafalp** [comment]
Made another pass on our codebase and updated todo's list accordingly. Now that `prices.discounts` are no longer a functions, and knowing what I know now about supporting currency conversions and with future tax computation in mind I'll need to rethink both how discounts are passed around the code as well as when `Money` becomes `TaxedMoney`.

**Comment by rafalp** [comment]
Two months in, this PR now introduces changes to over 104 files. I've rebased it against master and added dedicated document to our docs explaining our design decisions considering the money handling, currencies, `Money` vs `TaxedMoney` and `TaxedMoneyRange`.

Some things I'm unusure of:

- Second look on new GraphQL types as well as commit updating our JS to work with it is welcome
- Some things (eg. variable renames) are changed that are difficult to pick out in automated tests
- New documentation's contents

I'm also unsure of `get_gross_price()` method on `ShippingMethodCountry`. Its convenient but IMHO symptomatic of semantic issue that this change introduces to your codebase, where `price` is `Money` that then becomes `TaxedMoney`. But this PR is quite big already, and I'm weary of expanding it's scope even further.

All ~jabs~  constructive feedbacks are welcome.

File: docs/architecture/money.rst [hunk 1]
```diff
@@ -0,0 +1,33 @@
+Handling Money Amounts
+======================
+
+Saleor uses the `Prices <https://github.com/mirumee/prices/>`_ and `django-prices <https://github.com/mirumee/django-prices/>`_ libraries to store, calculate and display amounts of money, prices and ranges of those.
+
+Those libraries were developed at the Mirumee, home of the Saleor, with one eye towards its current and future requirements and thus reflect certain design decisions one must be aware of when working with money amounts:
```
Comment (patrys) [inline docs/architecture/money.rst]: This paragraph reads like a cheap advertisement. Let's drop it.

File: docs/architecture/money.rst [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+Default currency
+----------------
+
+By design all prices entered in the dashboard and stored in database are in single currency, specified in the `DEFAULT_CURRENCY` setting. Saleor is capable of displaying prices in user's currency, but scenario where store offers different items with prices in different currencies is not supported. Additionally, no conversion is performed on amounts used to calculate orders total prices or sent to payment providers.
```
Comment (patrys) [inline docs/architecture/money.rst]: Let's reword this:

All prices are entered and stored in a single default currency controlled by the [`DEFAULT_CURRENCY` setting]. Saleor can display prices in a user's local currency (see [Open Exchange Rates integration]) but all purchases are charged in the default currency.

File: docs/architecture/money.rst [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+There's nothing stopping you from changing that setting in an active store, but this will make amounts kept for historical reasons (eg. order items) inaccurate.
```
Comment (patrys) [inline docs/architecture/money.rst]: Let's reword:

.. warning::

The currency is not stored in the database. Changing the default currency in a production environment will not recalculate any existing orders. All numbers will remain the same and will be incorrectly displayed as the new currency.

File: saleor/cart/views.py [hunk 4]
```diff
@@ -126,8 +127,9 @@ def prepare_line_data(line):
             'attributes': line.variant.display_variant_attributes(attributes),
             'image': first_image,
             'price_per_item': currencyfmt(
-                price_per_item.gross, price_per_item.currency),
-            'line_total': currencyfmt(line_total.gross, line_total.currency),
+                price_per_item.gross.amount, price_per_item.currency),
+            'line_total': currencyfmt(
```
Comment (patrys) [inline saleor/cart/views.py]: Looks like we should create a convenience wrapper for `currencyfmt`:

```python
def format_money(money):
    return currencyfmt(money.amount, money.currency)
```

File: saleor/checkout/core.py [hunk 5]
```diff
@@ -210,32 +212,41 @@ def billing_address(self, address):
         self.modified = True
 
     @property
-    def discount(self):
-        """Return a discount if any."""
+    def discount_amount(self):
```
Comment (patrys) [inline saleor/checkout/core.py]: "Amount" suggests that it returns a number. `discount` was ok to keep, paired with the new `discount_name`.

File: saleor/checkout/core.py [hunk 6]
```diff
@@ -305,15 +316,19 @@ def create_order(self):
...
+                net=Money(0, currency=settings.DEFAULT_CURRENCY),
+                gross=Money(0, currency=settings.DEFAULT_CURRENCY))
+
         order_data = {
             'language_code': get_language(),
             'billing_address': billing_address,
             'shipping_address': shipping_address,
             'tracking_client_id': self.tracking_code,
-            'shipping_price': shipping_price,
+            'shipping_price': shipping_price.gross,
```
Comment (patrys) [inline saleor/checkout/core.py]: Why do we intentionally discard the net price here?

File: saleor/checkout/core.py [hunk 7]
```diff
@@ -368,32 +382,43 @@ def recalculate_discount(self):
...
-            del self.discount
+            del self.discount_amount
+            del self.discount_name
             del self.voucher_code
 
     def get_subtotal(self):
-        """Calculate order total without shipping."""
-        zero = Price(0, currency=settings.DEFAULT_CURRENCY)
+        """Calculate order total without shipping and discount."""
+        zero_subtotal = TaxedMoney(
```
Comment (patrys) [inline saleor/checkout/core.py]: Shouldn't this be a module-level constant?

File: saleor/checkout/forms.py [hunk 8]
```diff
@@ -68,7 +68,7 @@ class ShippingCountryChoiceField(forms.ModelChoiceField):
 
     def label_from_instance(self, obj):
         """Return a friendly label for the shipping method."""
-        price_html = format_price(obj.price.gross, obj.price.currency)
+        price_html = format_price(obj.price.amount, obj.price.currency)
```
Comment (patrys) [inline saleor/checkout/forms.py]: Why does it need two parameters? It looks like this should only accept a `Money` instance?

File: saleor/core/utils/__init__.py [hunk 9]
```diff
@@ -110,10 +110,10 @@ def get_user_shipping_country(request):
     return request.country
 
 
-def serialize_decimal(obj):
-    if isinstance(obj, decimal.Decimal):
-        return str(obj)
-    return JSONEncoder().default(obj)
+def serialize_amount(obj):
+    if isinstance(obj, Money):
+        return str(obj.amount)
```
Comment (patrys) [inline saleor/core/utils/__init__.py]: This looks wrong. If we only want the number to appear in JSON, we should only pass the amount to the serialized structure. This seems to be fixing data at the wrong level.

File: saleor/dashboard/order/forms.py [hunk 10]
```diff
@@ -60,13 +60,14 @@ def clean(self):
...
             None, pgettext_lazy(
-                'Payment form error',
-                'Payment gateway error: %s') % message)
+                'Payment form error', 'Payment gateway error: %s') % message)
 
     def try_payment_action(self, action):
-        amount = self.cleaned_data['amount']
+        # Unwrap Money() object to Decimal before passing it
+        # to payments utility library for processing
+        amount = self.cleaned_data['amount'].amount
```
Comment (patrys) [inline saleor/dashboard/order/forms.py]: I think this would be less confusing had it stayed in its previous form:

```python
money = self.cleaned_data['amount']
try:
    action(money.amount)
...
```

File: saleor/dashboard/product/views.py [hunk 11]
```diff
@@ -475,10 +475,16 @@ def variant_details(request, product_pk, variant_pk):
...
+        costs = {'min': costs_data['costs'][0],
+                 'max': costs_data['costs'][-1]}
+    else:
+        costs = {}
+
     ctx = {
         'images': images, 'product': product, 'stock': stock,
-        'variant': variant, 'costs': costs_data['costs'],
-        'margins': costs_data['margins'], 'is_empty': not stock.exists()}
+        'variant': variant, 'costs': costs, 'margins': costs_data['margins'],
```
Comment (patrys) [inline saleor/dashboard/product/views.py]: What if `costs_data` is empty? It seems possible given the check in line 478.

File: saleor/discount/forms.py [hunk 12]
```diff
@@ -42,15 +42,18 @@ def clean(self):
         if 'voucher' in cleaned_data:
             voucher = cleaned_data['voucher']
             try:
-                discount = get_voucher_discount_for_checkout(
+                discount_amount = get_voucher_discount_for_checkout(
```
Comment (patrys) [inline saleor/discount/forms.py]: Again, "amount" seems to indicate a number while this (I assume) returns a Money object.

File: saleor/discount/models.py [hunk 13]
```diff
@@ -109,31 +110,30 @@ def get_apply_to_display(self):
...
+            return partial(fixed_discount, discount=discount_amount)
+        if self.discount_value_type == DiscountValueType.PERCENTAGE:
+            return partial(percentage_discount, percentage=self.discount_value)
+        raise NotImplementedError('Unknown discount type')
+
+    def get_discount_amount_for(self, price):
+        discount = self.get_discount()
+        gross_amount = price.gross
+        gross_after_discount = discount(gross_amount)
+        if gross_after_discount < Money(0, currency=settings.DEFAULT_CURRENCY):
```
Comment (patrys) [inline saleor/discount/models.py]: `if gross_after_discount.amount < 0:`? Also wouldn't it be easier to make sure that discounts do not result in negative amounts?

File: saleor/discount/utils.py [hunk 14]
```diff
@@ -57,18 +56,17 @@ def get_product_discounts(product, discounts):
 def calculate_discounted_price(product, price, discounts):
     """Return minimum product's price of all prices with discounts applied."""
     if discounts:
-        discounts = list(
-            get_product_discounts(product, discounts))
+        discounts = list(get_product_discounts(product, discounts))
```
Comment (patrys) [inline saleor/discount/utils.py]: Do we need to cast it to a list here?

File: saleor/graphql/core/types.py [hunk 15]
```diff
@@ -33,20 +33,35 @@ class Error(graphene.ObjectType):
...
+    amount = graphene.Float()
+    localized = graphene.String()
+
+    def resolve_localized(self, info):
+        return prices_i18n.amount(self)
+
+
+class TaxedMoney(graphene.ObjectType):
     currency = graphene.String()
     gross = graphene.Float()
```
Comment (patrys) [inline saleor/graphql/core/types.py]: Why aren't `gross` and `net` fields of type `Money` you define above? This would also take core of currency and currency exchange.

File: saleor/order/utils.py [hunk 16]
```diff
@@ -42,16 +42,20 @@ def cancel_order(order):
...
-    total_gross = sum(p.gross for p in prices)
-    total = Price(
-        net=total_net, gross=total_gross, currency=settings.DEFAULT_CURRENCY)
-    total += order.shipping_price
+    total_net = sum(p.net.amount for p in prices)
+    total_gross = sum(p.gross.amount for p in prices)
+    total = TaxedMoney(
+        net=Money(total_net, currency=settings.DEFAULT_CURRENCY),
+        gross=Money(total_gross, currency=settings.DEFAULT_CURRENCY))
+    total += TaxedMoney(net=order.shipping_price, gross=order.shipping_price)
```
Comment (patrys) [inline saleor/order/utils.py]: Why is shipping price not a taxed amount already?

File: saleor/order/utils.py [hunk 17]
```diff
@@ -42,16 +42,20 @@ def cancel_order(order):
...
-    total_net = sum(p.net for p in prices)
-    total_gross = sum(p.gross for p in prices)
-    total = Price(
-        net=total_net, gross=total_gross, currency=settings.DEFAULT_CURRENCY)
-    total += order.shipping_price
+    total_net = sum(p.net.amount for p in prices)
+    total_gross = sum(p.gross.amount for p in prices)
+    total = TaxedMoney(
+        net=Money(total_net, currency=settings.DEFAULT_CURRENCY),
+        gross=Money(total_gross, currency=settings.DEFAULT_CURRENCY))
```
Comment (patrys) [inline saleor/order/utils.py]: Isn't this (and the previous lines) effectively `sum(prices, zero)`?

File: templates/checkout/details.html [hunk 18]
```diff
@@ -50,25 +50,27 @@ <h3 class="checkout__review__tittle">{% trans "Order review" context "Order revi
...
-              {% else %}
-                &mdash;
-              {% endif %}</p>
+              <p class="float-right">
+                {% if checkout.shipping_method %}
+                  {{ checkout.shipping_method.get_total_price.gross|amount }}
+                {% else %}
+                  &mdash;
+                {% endif %}
+              </p>
```
Comment (patrys) [inline templates/checkout/details.html]: Please verify that the reformatting did not affect how it's rendered.

File: templates/dashboard/discount/sale/list.html [hunk 19]
```diff
@@ -56,7 +56,7 @@
                       {{ sale }}
                     </td>
                     <td>
-                      {{ sale.value }}{{ sale.get_type_display }}
+                      {{ sale.value }}&nbsp;<span class="currency">{{ sale.get_type_display }}</span>
```
Comment (patrys) [inline templates/dashboard/discount/sale/list.html]: This looks like an unrelated change?

File: saleor/cart/views.py [hunk 20]
```diff
@@ -93,16 +93,17 @@ def update(request, cart, variant_id):
...
             response['subtotal'] = currencyfmt(
-                updated_line.get_total(discounts=discounts).gross,
+                updated_line.get_total(discounts=discounts).gross.amount,
                 updated_line.get_total(discounts=discounts).currency)
         if cart:
             cart_total = cart.get_total(discounts=discounts)
             response['total'] = currencyfmt(
-                cart_total.gross, cart_total.currency)
+                cart_total.gross.amount, cart_total.currency)
+            # FIXME: Currency exchange doesn't work with new prices package
```
Comment (patrys) [inline saleor/cart/views.py]: This is a blocker if true.

File: saleor/order/utils.py [hunk 21]
```diff
@@ -42,16 +42,20 @@ def cancel_order(order):
...
+    total += TaxedMoney(net=order.shipping_price, gross=order.shipping_price)
+    if order.discount_amount:
+        total -= order.discount_amount
```
Comment (mad-anne) [inline saleor/order/utils.py]: We update here total decreasing total by discount_amount. But what if order has applied voucher "5 USD discount for each book" and staff user has just increased quantity of ordered books?

Comment (patrys) [reply saleor/order/utils.py]: I think it's safer if we allowed staff to manually adjust prices after modifying an order. The original discount could no longer exist, could no longer apply (because of date ranges, because of usage limits, because a qualifying row was removed from the order etc.) and so on.

Comment (rafalp) [reply saleor/order/utils.py]: Recounting the orders based on current or past state of the store configuration is something that I've been discussing with @elwoodxblues for some time now. Last time we've talked about this it felt to us that no mechanic may be smarter here than human modifying the order and thus it should be up to human handling the editing to be able to change discount's amount based on circumstances.

File: templates/checkout/details.html (hunk 18 above)
Comment (rafalp) [reply templates/checkout/details.html]: Looking same

File: saleor/discount/utils.py (hunk 14 above)
Comment (rafalp) [reply saleor/discount/utils.py]: Without cast to list the `if discounts:` in next line will get generator and always evaluate to true and thus raise `min() arg is an empty sequence` if nothing is returned.

File: saleor/discount/models.py (hunk 13 above)
Comment (rafalp) [reply saleor/discount/models.py]: Yeah, comparing amount against zero is simpler than what I've came up with.

As for making sure that discount won't be negative, I've experimented with being safe before being sorry here, but couldn't come with much. For fixed discounts, you can check for `price.gross.amount <= self.discount_value`, and for percentage for `self.discount_value > 100`. For percentage discounts we could add check to dashboard's form to prevent nonsensical discounts like 101%. On sidenote, checking for nergative has its upsides too, custom discount types in forks would be guaranteed to never produce negatives without having to modify the check. Just adding custom type to `get_discount` would cut it.

Comment (patrys) [reply saleor/discount/models.py]: I meant that the actual implementations of `fixed_discount` and `percentage_discount` should never return a negative number. Otherwise all consumers of the `prices` API have to add these checks in their code.

File: saleor/dashboard/product/views.py (hunk 11 above)
Comment (rafalp) [reply saleor/dashboard/product/views.py]: `get_variant_costs_data()` return dict is guaranteed to always contain `costs` and `margins` keys, even if those are empty lists, so fixed condition to `if costs_data['costs']`

Comment (patrys) [reply saleor/dashboard/product/views.py]: Wouldn't it be more useful if it returned a named tuple or a `__slots__`-based object instead?
